File: frontend/src/lib/detector.js
// 1:1 JS port of backend/server.py segment-effort detection.
// Uses the 30m haversine radius tolerance, matches start then end in order,
// and computes elapsed time / average power / average HR per effort.

export const DETECTION_RADIUS_M = 30.0;
export const MAX_DISPLAY_POINTS = 1500;

export function haversineM(lat1, lon1, lat2, lon2) {
  const R = 6371000.0;
  const p1 = (lat1 * Math.PI) / 180;
  const p2 = (lat2 * Math.PI) / 180;
  const dp = ((lat2 - lat1) * Math.PI) / 180;
  const dl = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dp / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) ** 2;
  return 2 * R * Math.asin(Math.sqrt(a));
}

export function totalDistanceM(points) {
  let d = 0.0;
  for (let i = 1; i < points.length; i++) {
    d += haversineM(points[i - 1].lat, points[i - 1].lon, points[i].lat, points[i].lon);
  }
  return d;
}

export function elevationGainM(points) {
  let gain = 0.0;
  for (let i = 1; i < points.length; i++) {
    const e1 = points[i - 1].ele;
    const e2 = points[i].ele;
    if (e1 != null && e2 != null && e2 > e1) gain += e2 - e1;
  }
  return gain;
}

export function decimate(points, maxPoints = MAX_DISPLAY_POINTS) {
  if (points.length <= maxPoints) return points;
  const step = Math.max(1, Math.floor(points.length / maxPoints));
  const out = [];
  for (let i = 0; i < points.length; i += step) out.push(points[i]);
  if (out[out.length - 1] !== points[points.length - 1]) {
    out.push(points[points.length - 1]);
  }
  return out;
}

export function detectEfforts(ridePoints, segment) {
  const segPts = segment.points;
  if (segPts.length < 2 || !ridePoints.length) return [];
  const segStart = segPts[0];
  const segEnd = segPts[segPts.length - 1];

  const efforts = [];
  let i = 0;
  const n = ridePoints.length;

  while (i < n) {
    // find entry into start radius
    let startIdx = null;
    let bestStartDist = null;
    while (i < n) {
      const d = haversineM(
        ridePoints[i].lat, ridePoints[i].lon, segStart.lat, segStart.lon
      );
      if (d <= DETECTION_RADIUS_M) {
        startIdx = i;
        bestStartDist = d;
        let j = i + 1;
        while (j < n) {
          const dj = haversineM(
            ridePoints[j].lat, ridePoints[j].lon, segStart.lat, segStart.lon
          );
          if (dj <= DETECTION_RADIUS_M && dj < bestStartDist) {
            bestStartDist = dj;
            startIdx = j;
            j += 1;
          } else {
            break;
          }
        }
        i = j;
        break;
      }
      i += 1;
    }
    if (startIdx === null) break;

    // find exit to end point
    let endIdx = null;
    let bestEndDist = null;
    let k = i;
    while (k < n) {
      const d = haversineM(
        ridePoints[k].lat, ridePoints[k].lon, segEnd.lat, segEnd.lon
      );
      if (d <= DETECTION_RADIUS_M) {
        endIdx = k;
        bestEndDist = d;
        let m = k + 1;
        while (m < n) {
          const dm = haversineM(
            ridePoints[m].lat, ridePoints[m].lon, segEnd.lat, segEnd.lon
          );
          if (dm <= DETECTION_RADIUS_M && dm < bestEndDist) {
            bestEndDist = dm;
            endIdx = m;
            m += 1;
          } else {
            break;
          }
        }
        k = m;
        break;
      }
      k += 1;
    }
    if (endIdx === null || endIdx <= startIdx) break;

    // compute effort metrics
    const slice = ridePoints.slice(startIdx, endIdx + 1);
    const tStart = slice[0].t;
    const tEnd = slice[slice.length - 1].t;
    if (!tStart || !tEnd) {
      i = endIdx + 1;
      continue;
    }
    const dtStart = new Date(tStart);
    const dtEnd = new Date(tEnd);
    if (Number.isNaN(dtStart.getTime()) || Number.isNaN(dtEnd.getTime())) {
      i = endIdx + 1;
      continue;
    }
    const elapsed = (dtEnd.getTime() - dtStart.getTime()) / 1000;
    if (elapsed <= 0) {
      i = endIdx + 1;
      continue;
    }

    const powers = slice.filter((p) => p.power != null).map((p) => p.power);
    const hrs = slice.filter((p) => p.hr != null).map((p) => p.hr);
    const cads = slice.filter((p) => p.cad != null).map((p) => p.cad);
    const speeds = slice.filter((p) => p.speed != null).map((p) => p.speed);
    const avgPower = powers.length
      ? Math.round((powers.reduce((a, b) => a + b, 0) / powers.length) * 10) / 10
      : null;
    const avgHr = hrs.length
      ? Math.round((hrs.reduce((a, b) => a + b, 0) / hrs.length) * 10) / 10
      : null;
    const avgCad = cads.length
      ? Math.round((cads.reduce((a, b) => a + b, 0) / cads.length) * 10) / 10
      : null;
    const maxPower = powers.length ? Math.max(...powers) : null;
    const maxHr = hrs.length ? Math.max(...hrs) : null;
    const maxCad = cads.length ? Math.max(...cads) : null;

    const distance = totalDistanceM(slice);
    const eleGain = elevationGainM(slice);
    const avgSpeed = elapsed > 0 ? distance / elapsed : null;
    const maxSpeed = speeds.length
      ? Math.max(...speeds)
      : avgSpeed; // fallback

    efforts.push({
      segment_id: segment.id,
      start_idx: startIdx,
      end_idx: endIdx,
      elapsed_s: elapsed,
      moving_time_s: elapsed, // GPX/FIT records have already been filtered
      distance_m: Math.round(distance * 10) / 10,
      elevation_gain_m: Math.round(eleGain * 10) / 10,
      avg_speed_mps: avgSpeed != null ? Math.round(avgSpeed * 100) / 100 : null,
      max_speed_mps: maxSpeed != null ? Math.round(maxSpeed * 100) / 100 : null,
      avg_power: avgPower,
      max_power: maxPower,
      avg_hr: avgHr,
      max_hr: maxHr,
      avg_cadence: avgCad,
      max_cadence: maxCad,
      datetime_utc: dtStart.toISOString(),
    });
    i = endIdx + 1;
  }
  return efforts;
}

// ---------- Hashing (SHA-256 via WebCrypto) ----------
async function sha256Hex(str) {
  const enc = new TextEncoder().encode(str);
  const buf = await crypto.subtle.digest("SHA-256", enc);
  return Array.from(new Uint8Array(buf))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

export async function hashSegment(points) {
  const key = points
    .map((p) => `${p.lat.toFixed(5)},${p.lon.toFixed(5)}`)
    .join(";");
  return sha256Hex(key);
}

export async function hashRide(points) {
  const step = Math.max(1, Math.floor(points.length / 200) || 1);
  const parts = [];
  for (let i = 0; i < points.length; i += step) {
    const p = points[i];
    parts.push(`${p.lat.toFixed(5)},${p.lon.toFixed(5)},${p.t || ""}`);
  }
  return sha256Hex(parts.join(";"));
}

File: backend/cst/detector.py
"""Geometry + segment-effort matching ported from the previous JS implementation.

This is a 1:1 port of `frontend/src/lib/detector.js` (including the hysteresis
elevation-gain algorithm we tuned in Pass 1 of the previous session).
"""

from __future__ import annotations

import math
from typing import Iterable, List, Sequence

EARTH_R_M = 6371000.0
DETECTION_RADIUS_M = 30.0
MAX_DISPLAY_POINTS = 2000
ELE_THRESHOLD_M = 3.0


def haversine_m(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    a = math.radians(lat1)
    b = math.radians(lat2)
    dlat = b - a
    dlon = math.radians(lon2 - lon1)
    s = math.sin(dlat / 2) ** 2 + math.cos(a) * math.cos(b) * math.sin(dlon / 2) ** 2
    return 2 * EARTH_R_M * math.asin(min(1.0, math.sqrt(s)))


def total_distance_m(points: Sequence[dict]) -> float:
    d = 0.0
    for i in range(1, len(points)):
        p, q = points[i - 1], points[i]
        d += haversine_m(p["lat"], p["lon"], q["lat"], q["lon"])
    return d


def elevation_gain_m(points: Sequence[dict], threshold: float = ELE_THRESHOLD_M) -> float:
    gain = 0.0
    base = None
    peak = None
    for p in points:
        e = p.get("ele")
        if e is None:
            continue
        if base is None:
            base = peak = e
            continue
        if e > peak:
            peak = e
        elif peak - e >= threshold:
            if peak > base:
                gain += peak - base
            base = peak = e
        elif e < base:
            base = peak = e
    if base is not None and peak is not None and peak > base:
        gain += peak - base
    return gain


def elevation_loss_m(points: Sequence[dict], threshold: float = ELE_THRESHOLD_M) -> float:
    loss = 0.0
    crest = None
    trough = None
    for p in points:
        e = p.get("ele")
        if e is None:
            continue
        if crest is None:
            crest = trough = e
            continue
        if e < trough:
            trough = e
        elif e - trough >= threshold:
            if crest > trough:
                loss += crest - trough
            crest = trough = e
        elif e > crest:
            crest = trough = e
    if crest is not None and trough is not None and crest > trough:
        loss += crest - trough
    return loss


def decimate(points: Sequence[dict], max_points: int = MAX_DISPLAY_POINTS) -> List[dict]:
    """Reduce a point series to at most `max_points` for transport/display."""
    if len(points) <= max_points:
        return list(points)
    step = max(1, len(points) // max_points)
    out = list(points[::step])
    if out[-1] is not points[-1]:
        out.append(points[-1])
    return out


def _closest_index(points: Sequence[dict], target: dict) -> tuple[int, float]:
    """Index + distance of the point closest to `target` (linear scan)."""
    best_i = -1
    best_d = float("inf")
    tlat, tlon = target["lat"], target["lon"]
    for i, p in enumerate(points):
        d = haversine_m(p["lat"], p["lon"], tlat, tlon)
        if d < best_d:
            best_d = d
            best_i = i
    return best_i, best_d


def _cluster_indices(indices: Sequence[int], max_gap: int = 8) -> List[List[int]]:
    """Group a sorted list of ride-point indices into contiguous clusters.

    A cluster is a run of indices whose gap to the previous one is ≤ max_gap.
    Used to collapse the dozen-or-so consecutive in-disc points that one
    physical pass produces into a single logical entry/exit event. The 8-
    point default tolerates ~8 s of GPS dropout / under-bridge noise without
    splitting a single physical pass into two phantom passes, while still
    leaving real out-and-back loops as separate clusters (the rider has to
    leave the disc for many more seconds to come back).
    """
    out: List[List[int]] = []
    for i in indices:
        if out and i - out[-1][-1] <= max_gap:
            out[-1].append(i)
        else:
            out.append([i])
    return out


def detect_efforts(
    ride_points: Sequence[dict],
    segment_points: Sequence[dict],
    radius: float = DETECTION_RADIUS_M,
) -> List[dict]:
    """Find every contiguous slice of `ride_points` that starts within `radius`
    of the segment start, passes through the middle, and ends within `radius`
    of the segment end. Returns a list of effort summaries.

    The matcher clusters consecutive in-disc points so that a single physical
    pass (which yields ~12 GPS points inside the 30 m disc at typical cycling
    speeds) becomes one logical "start" + one logical "end" event, not 12.
    """
    if len(ride_points) < 2 or len(segment_points) < 2:
        return []
    seg_start = segment_points[0]
    seg_end = segment_points[-1]

    # Compute distance to seg_start / seg_end for every ride point once.
    start_disc_idxs: List[int] = []
    end_disc_idxs: List[int] = []
    for i, p in enumerate(ride_points):
        if haversine_m(p["lat"], p["lon"], seg_start["lat"], seg_start["lon"]) <= radius:
            start_disc_idxs.append(i)
        if haversine_m(p["lat"], p["lon"], seg_end["lat"], seg_end["lon"]) <= radius:
            end_disc_idxs.append(i)
    if not start_disc_idxs or not end_disc_idxs:
        return []

    # Collapse runs of consecutive in-disc points (one physical pass = one cluster).
    # Inside each cluster we keep the index closest to the segment endpoint —
    # that yields the tightest slice timing.
    def _best_in_cluster(cluster: List[int], target: dict) -> int:
        best_i = cluster[0]
        best_d = float("inf")
        for i in cluster:
            p = ride_points[i]
            d = haversine_m(p["lat"], p["lon"], target["lat"], target["lon"])
            if d < best_d:
                best_d = d
                best_i = i
        return best_i

    start_clusters = _cluster_indices(start_disc_idxs)
    end_clusters = _cluster_indices(end_disc_idxs)
    starts = [_best_in_cluster(c, seg_start) for c in start_clusters]
    ends = [_best_in_cluster(c, seg_end) for c in end_clusters]

    used_end = -1
    efforts: List[dict] = []
    mid = segment_points[len(segment_points) // 2]
    seg_distance = total_distance_m(segment_points)
    min_slice_distance = seg_distance * 0.6  # reject slices that obviously didn't cover the segment
    for s in starts:
        # A start that falls inside an already-consumed slice is a duplicate
        # caused by GPS jitter splitting a single physical entry into two
        # nearby start clusters — skip it.
        if s <= used_end:
            continue
        e = next((j for j in ends if j > s and j > used_end), None)
        if e is None:
            continue
        slice_pts = ride_points[s : e + 1]
        # Sanity 1: must pass near the segment midpoint — guards against
        # spurious matches where the rider entered the start disc but
        # didn't actually traverse the segment (e.g. they crossed a road
        # tangentially, or did a U-turn).
        _, mid_d = _closest_index(slice_pts, mid)
        if mid_d > radius * 4:
            continue
        # Sanity 2: the slice must cover at least 60% of the segment's own
        # length. This filters out the "GPS jitter created a fake second
        # cluster a few metres from the first" failure mode that produced
        # duplicate near-identical efforts in earlier builds.
        if total_distance_m(slice_pts) < min_slice_distance:
            continue
        used_end = e
        efforts.append(_summarise_effort(slice_pts, s, e))
    return efforts


def _summarise_effort(slice_pts: Sequence[dict], s: int, e: int) -> dict:
    t0 = slice_pts[0].get("t")
    t1 = slice_pts[-1].get("t")
    elapsed = 0.0
    if t0 and t1:
        from datetime import datetime

        try:
            dt0 = datetime.fromisoformat(t0.replace("Z", "+00:00"))
            dt1 = datetime.fromisoformat(t1.replace("Z", "+00:00"))
            elapsed = (dt1 - dt0).total_seconds()
        except Exception:
            elapsed = 0.0

    def _avg(key: str) -> float | None:
        vals = [p[key] for p in slice_pts if p.get(key) is not None]
        return sum(vals) / len(vals) if vals else None

    def _max(key: str) -> float | None:
        vals = [p[key] for p in slice_pts if p.get(key) is not None]
        return max(vals) if vals else None

    # Moving time = elapsed minus long stationary gaps (>5 s apart).
    moving = elapsed
    if t0 and t1 and len(slice_pts) > 2:
        from datetime import datetime

        try:
            stops = 0.0
            for i in range(1, len(slice_pts)):
                a = slice_pts[i - 1].get("t")
                b = slice_pts[i].get("t")
                if not a or not b:
                    continue
                da = datetime.fromisoformat(a.replace("Z", "+00:00"))
                db = datetime.fromisoformat(b.replace("Z", "+00:00"))
                dt = (db - da).total_seconds()
                if dt > 5:
                    dx = haversine_m(
                        slice_pts[i - 1]["lat"],
                        slice_pts[i - 1]["lon"],
                        slice_pts[i]["lat"],
                        slice_pts[i]["lon"],
                    )
                    if dx < 2:
                        stops += dt
            moving = max(0.0, elapsed - stops)
        except Exception:
            pass

    distance = total_distance_m(slice_pts)
    avg_speed = (distance / moving) if moving > 0 else None
    max_speed_pt = _max("speed")
    if max_speed_pt is None:
        # derive from consecutive points
        m = 0.0
        from datetime import datetime

        for i in range(1, len(slice_pts)):
            a = slice_pts[i - 1]
            b = slice_pts[i]
            if not a.get("t") or not b.get("t"):
                continue
            try:
                da = datetime.fromisoformat(a["t"].replace("Z", "+00:00"))
                db = datetime.fromisoformat(b["t"].replace("Z", "+00:00"))
                dt = (db - da).total_seconds()
                if dt <= 0:
                    continue
                v = haversine_m(a["lat"], a["lon"], b["lat"], b["lon"]) / dt
                if v > m:
                    m = v
            except Exception:
                continue
        max_speed_pt = m if m > 0 else None

    return {
        "start_idx": s,
        "end_idx": e,
        "elapsed_s": elapsed,
        "moving_time_s": moving,
        "distance_m": distance,
        "avg_power": _avg("power"),
        "max_power": _max("power"),
        "avg_hr": _avg("hr"),
        "max_hr": _max("hr"),
        "avg_cadence": _avg("cad"),
        "avg_speed_mps": avg_speed,
        "max_speed_mps": max_speed_pt,
        "elevation_gain_m": elevation_gain_m(slice_pts),
        "datetime_utc": t0,
    }


def hash_segment_points(points: Iterable[dict]) -> str:
    """Stable hash of segment geometry (rounded coordinates)."""
    import hashlib

    h = hashlib.sha256()
    for p in points:
        h.update(f"{round(p['lat'], 6)},{round(p['lon'], 6)};".encode())
    return h.hexdigest()


def hash_ride_points(points: Iterable[dict]) -> str:
    """Stable hash of a ride: start coordinate + time + point count."""
    import hashlib

    pts = list(points)
    h = hashlib.sha256()
    if pts:
        first = pts[0]
        last = pts[-1]
        h.update(
            f"{round(first['lat'], 5)},{round(first['lon'], 5)},{first.get('t','')},"
            f"{round(last['lat'], 5)},{round(last['lon'], 5)},{last.get('t','')},"
            f"{len(pts)}".encode()
        )
    return h.hexdigest()
